### ISSUE TITLE
508-defect-2 [FIELD TEMPLATE]: Description and heading can separate label from input, causing strange screen reader experiences

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#10095]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Description

The BAM2 team [working on the medical device reordering form 2346](https://github.com/department-of-veterans-affairs/va.gov-team/issues/9860) found a strange issue with the Email Widget. Essentially, the input label was being rendered as a heading, and was separated from the input. This lead to a situation where users could click on the `<h4>` heading and the input would receive focus. It also could be confusing to screen reader users.

After some sleuthing, I determined the root cause was our [Field Template component](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/components/FieldTemplate.jsx) and how we're building the label and description.

## Related Issues
* https://github.com/department-of-veterans-affairs/va.gov-team/issues/9860

## Point of Contact

**VFS Point of Contact:** _Trevor_

## Acceptance Criteria

- [ ] The Field Template component will allow users to enter an optional heading and description, before the label + input.
- [ ] The heading ideally will allow users to set which level (2, 3, 4, etc) so we can avoid extra heading levels should increase by one warnings.
- [ ] No axe errors should be thrown when we have a heading, description, label, input, or some combination thereof.

## Steps to Recreate

1. Enter https://staging.va.gov/hearing-aid-batteries-and-accessories/veteran-information/addresses while logged in as user 1 in browser
1. Inspect the Email address code
1. Verify that the code appears as below in the "Current code"

## Possible Fixes (optional)

I made some adjustments to the template, shown here as diff additions:

```diff
+ const testHeader = uiSchema['ui:header'];
+ const textHeader = typeof testHeader === 'string' ? testHeader : null;
```

```diff
// FieldTemplate.jsx#L103

const content = (
<>
+   {textHeader && <h4>{textHeader}</h4>}
!   // Text description should come after the heading, before the label
!   {textDescription && <p>{textDescription}</p>}
    {showLabel && labelElement}
    {DescriptionField && (
      <DescriptionField options={uiSchema['ui:options']} />
    )}
    {!textDescription && !DescriptionField && description}
    {errorSpan}
    {<div className={inputWrapperClassNames}>{children}</div>}
    {help}
  </>
);
```

## WCAG or Vendor Guidance (optional)

* [Meaningful Sequence: Understanding SC 1.3.2](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-sequence.html)

## Screenshots or Trace Logs

![Screen Shot 2020-06-04 at 5 12 54 PM](https://user-images.githubusercontent.com/57469/83811256-77449300-a687-11ea-9785-6840db1115d5.png)

**Comment by Clark-W-Griswold** [comment]
Looks like this might be a form configuration issue. I have discovered other forms which use a different pattern to display this same UI. In these cases the HTML produced will contain the desired "heading, description, label, input" pattern suggested. 

![image.png](https://images.zenhubusercontent.com/5db31b5573e8f70001d9e0c8/3a66773a-dff5-4c8a-989b-15f317c5f275)

**Comment by Clark-W-Griswold** [comment]
Url for above: https://staging.va.gov/education/apply-for-education-benefits/application/1995/personal-information/contact-information

**Comment by bkjohnson** [comment]
I couldn't get the original link that Trevor posted to work, but I think Dan may be right.

![image](https://user-images.githubusercontent.com/2008881/90560745-6f5ba180-e154-11ea-9bef-fc11d70f604b.png)

I'm going to see if it is possible to prevent that alternative configuration from happening.

**Comment by bkjohnson** [comment]
What we're seeing in the markup in Trevor's is the result of the `ui:description` part of the schema being used to include the label as well.

https://github.com/department-of-veterans-affairs/vets-website/blob/15f0403d9b605c6108730249738024af74777991/src/applications/disability-benefits/2346/schemas/2346UI.js#L21

Here is where that configuration takes effect in the markup:

https://github.com/department-of-veterans-affairs/vets-website/blob/15f0403d9b605c6108730249738024af74777991/src/platform/forms-system/src/js/components/FieldTemplate.jsx#L107-L109

I took a glance to see how `ui:description` is used elsewhere and most apps either use a plain string or a simple piece of JSX with a `<p>` or `<div>` that _doesn't_ include anything meant to resemble a label.

**Comment by 1Copenut** [comment]
Dan asked me to take a look at his changes in staging behind a feature flag, using the form https://staging.va.gov/education/apply-for-education-benefits/application/1995/personal-information/contact-information

I reviewed it with JAWS, NVDA, and VoiceOver. This change is making the form labels behave correctly, and validates axe checks with 0 violations.

I'd like to see the `<legend>` not wrapped in a couple layers of nested `<div>` containers; some screen readers don't read the legend text when users are tabbing because of this. I believe Brooks is working on the legend in div issue with https://github.com/department-of-veterans-affairs/vets-website/pull/13321 .

This is a solid pattern that could be applied for the related issue #9860 too.

---
![Screen Shot 2020-08-19 at 9 23 34 AM](https://user-images.githubusercontent.com/934879/90647406-e7ca6d00-e1fd-11ea-9937-e96783ebf568.png)

**Comment by 1Copenut** [comment]
@bkjohnson && @Clark-W-Griswold It seems in conversations this wasn't a code issue, but more of a configuration issue? If that's accurate, I'm okay closing this ticket anytime.

**Comment by meganhkelley** [comment]
@bkjohnson @cvalarida is this still something y'all are working on or should we close?

**Comment by bkjohnson** [comment]
I'm working on it now that the defect-1s seem to be under control, but I still think that it is a duplicate of #9860

**Comment by bkjohnson** [comment]
Closing this in favor of #9860 since it appears to be a duplicate. Feel free to re-open if this isn't the case.
